### PR TITLE
`ProcessInstanceAssert.hasCalledProcess` doesn't check if the process contains the passed process id

### PR DESCRIPTION
**Pull request by Wolgo**
Actually check if a process with the correct processID is present

## Description

Improve function to actually test for correct process ID. 

## Related issues

closes #

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [ ] Javadoc has been written
* [ ] The documentation is updated

### CODE REVIEW

**Comment by Wolgo** [comment]
@remcowesterhoud I'm rudely tagging you, as I am otherwise going to forget about this, and it  Seems like a pretty important one, as this might result in tests incorrectly showing something calls the right subprocess.  Seems to me like this function was just forgetting to actually test for the correct ProcessId.

Didn't update tests or anything yet, because I have some other priorities right now.

**Review by remcowesterhoud — COMMENTED** [review]
Thanks for the bugfix @Wolgo! Your changes look great, I like how you kept the old method available so we can stay backwards compatible 👍 

As you mentioned you haven't written tests. I can't merge this without tests, but I do want to see this merged. We can do 2 things here. If you are interested in writing the tests still you can commit this and I will merge this PR. Alternatively, I can create an issue and branch for this bug. I can merge it to this branch and write some tests for this myself. What do you prefer?

**Comment by Wolgo** [comment]
`  public ProcessInstanceAssert findFirstCalledSubProcess(ProcessInstanceEvent ProcessEvent) {
    return new ProcessInstanceAssert(
        (StreamFilter.processInstance(this.recordStream)
            .withParentProcessInstanceKey(ProcessEvent.getProcessInstanceKey()).stream()
            .findFirst().orElseThrow(NoSuchElementException::new)).getKey(), this.recordStream);
  }`
  
  I quickly wrote this to work around that, and I'm thinking of just making it a generic function where you can get xth calls of process y.

**Comment by remcowesterhoud** [comment]
> @remcowesterhoud I think it's good If I write some tests, do you have any examples? I saw these classes were not really doing any useful tests yet. I don't know if we wanna juggle with a mock setup or just use a real process.

It would be something along the lines of this in the `AbstractProcessInstanceAssertTest`:
```java
    @Test
    void testHasCalledProcess() throws InterruptedException, TimeoutException {
      // given
      Utilities.deployResources(
          client,
          ProcessPackCallActivity.RESOURCE_NAME,
          ProcessPackCallActivity.CALLED_RESOURCE_NAME);

      // when
      final ProcessInstanceEvent instanceEvent =
          Utilities.startProcessInstance(engine, client, ProcessPackCallActivity.PROCESS_ID);

      // then
      BpmnAssert.assertThat(instanceEvent)
          .hasCalledProcess(ProcessPackCallActivity.CALLED_PROCESS_ID);
    }
```

Of course this is only a happy path test for one of the functions. It's also not a great test as this is the only call activity that's in this process. So I would expect we need to introduce a `ProcessPackMultipleCallActivities` which contains a process with multiple call activities with different ids.

> I'm running to some other issues with subprocesses, so I'll probably extend it a bit more. (Multi-instances always seem to do X+1 calls, meaning that ExtractingLastCalledProcess becomes useless in the case of multi-instances.)

Would you mind separating this in a different branch? That way we can keep this PR isolated to one issue at a time.

**Comment by remcowesterhoud** [comment]
Hi @Wolgo, hope you're doing well! This PR has been stale for a while, are you still interested in finishing this?

**Comment by Wolgo** [comment]
Hey @remcowesterhoud , doing well, between a programming job and trying and succeeding a house, I never ended up finishing this.
Luckily you caught me on day where I could 0 coding and only did paperwork, so was pretty excited for this.
I wrote some tests for this, they are a bit wierd, as we can't check the processId of the current process in scope, so instead we use the fact that a process has or has not called a process to differentiate. (To me that's a bit easier to keep in mind, as that's already in scope, you could look at specific elements passed, but then you have to hold two concepts in mind.)

I am open for suggestions on alternatives that make more sense here. 
I added a small line about multi-instance loops, that's not a bug of this test code but maybe of camunda. Can't really do anything about that here, but it's a good warning for people :). 

I added a "HasNotCalledProcess" as that seems convenient in general and was really helpful for tests.

File: qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java
Comment (remcowesterhoud) [inline qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:253]: ❌ Instead of introducing a `ProcessPackAlternateStartEndEvent` let's just use the `ProcessPackStartEndEvent`. This means you can reference the resource from the `ProcessPackStartEndEvent`. For the process id you'll need some unique string, e.g. `calledProcessIdOne`.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
> between a programming job and trying and succeeding to buy a house,

Sounds exciting!

Thanks for the changes you've made. Very thorough with the testing, I like it 🚀 

I have some minor suggestions. Mostly they are just some terminology fixes, but there's a few other changes as well. Please have a look and let me know if there's any questions!

Once you're done with the changes please run `mvn spotless:apply`. The CI is failing because of some code style issues 🙂

File: qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java
Comment (Wolgo) [reply qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:253]: I need a different processId between One and Three, to show a hasCalled(OneId) but hasNotCalled(ThreeId) on the parent project. 

Some of the requirements for my process that I have in mind: 
- Two sequential call activities, to show it correctly extracts the last called process.
- A call activity that has a call activity, to show the parent process does not detect nested child processes. (Which also means that the nested child process needs to have a different process Id than the 2 non-nested child processes.)

The process id is baked into the resource, so I made a new simple resource to prevent having to do some extra step.
Is there a way to deploy the same process twice with a different process id? 
Or is there another process that we can use for this?

**Comment by Wolgo** [comment]
I'll finish the rest this evening, my pipeline is done, so back to actual work

File: qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java
Comment (remcowesterhoud) [reply qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:253]: Oh sorry, my bad. I got confused with how we create processes in the Zeebe repo. You are totally right that you need the second process and resource. You can disregard these comments about removing the resource.

File: qa/abstracts/src/main/resources/alternate-start-end.bpmn
Comment (remcowesterhoud) [reply qa/abstracts/src/main/resources/alternate-start-end.bpmn:26]: Please ignore

File: qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java
Comment (remcowesterhoud) [reply qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:278]: Please ignore

File: assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java
Comment (Wolgo) [reply assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java]: I didn't see a way to count how many times each process is called, and if you are not intending to call a process, being told which distinct processes are called feels pretty good already.
But agreed, so implemented.

File: assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java
Comment (remcowesterhoud) [inline assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java]: 🔧 One more minor change before I can merge it. We already get the caller process records on line 634. Please store the result of this in a variable instead of getting them again here. E.g.:

```java
  public ProcessInstanceAssert hasNotCalledProcess() {
    final Set<String> calledProcesses =
        getCalledProcessRecords().stream()
            .map(process -> process.getValue().getBpmnProcessId())
            .collect(Collectors.toSet());

    assertThat(!calledProcesses.isEmpty())
        .withFailMessage(
            "A process was called from this process, distinct called processes are: %s",
            calledProcesses)
        .isFalse();
    return this;
  }
```

You will have to swap the processes in the failure message of your unit test.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
🚀 Great job, there is 1 more minor thing. After this I can merge the PR 🎉

File: assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java
Comment (Wolgo) [reply assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java]: Ah, Collectors.toSet() removes the need for distinct. I still gotta get more used to streams in Java. 
Thanks for the reviewing!

**Review by remcowesterhoud — APPROVED** [review]
All good! Thanks for the contribution 🏆 And good luck with the new job and house 🙂

**Pull request merged**
